File: backend/app/api/endpoints/locations.py
from typing import Any, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app import crud, models, schemas
from app.api import deps

router = APIRouter()


@router.get("/", response_model=List[schemas.Location])
def read_locations(
    db: Session = Depends(deps.get_db),
    skip: int = 0,
    limit: int = 100,
    parent_id: Optional[int] = None,
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Retrieve locations.
    """
    if parent_id is not None:
        locations = crud.location.get_multi_by_parent(
            db, parent_id=parent_id, owner_id=current_user.id
        )
    else:
        locations = crud.location.get_multi_by_owner(
            db, owner_id=current_user.id, skip=skip, limit=limit
        )
    return locations


@router.get("/tree", response_model=List[schemas.LocationTree])
def read_location_tree(
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Retrieve location tree.
    """
    return crud.location.get_location_tree(db, owner_id=current_user.id)


@router.post("/", response_model=schemas.Location)
def create_location(
    *,
    db: Session = Depends(deps.get_db),
    location_in: schemas.LocationCreate,
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Create new location.
    """
    location = crud.location.create(db=db, obj_in=location_in, owner_id=current_user.id)
    return location


@router.get("/{id}", response_model=schemas.Location)
def read_location(
    *,
    db: Session = Depends(deps.get_db),
    id: int,
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Get location by ID.
    """
    location = crud.location.get(db=db, id=id)
    if not location:
        raise HTTPException(status_code=404, detail="Location not found")
    if location.owner_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    return location


@router.put("/{id}", response_model=schemas.Location)
def update_location(
    *,
    db: Session = Depends(deps.get_db),
    id: int,
    location_in: schemas.LocationUpdate,
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Update a location.
    """
    location = crud.location.get(db=db, id=id)
    if not location:
        raise HTTPException(status_code=404, detail="Location not found")
    if location.owner_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    location = crud.location.update(db=db, db_obj=location, obj_in=location_in)
    return location


@router.delete("/{id}", response_model=schemas.Location)
def delete_location(
    *,
    db: Session = Depends(deps.get_db),
    id: int,
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Delete a location.
    """
    location = crud.location.get(db=db, id=id)
    if not location:
        raise HTTPException(status_code=404, detail="Location not found")
    if location.owner_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    
    # Check if location has items
    items = crud.item.get_by_location(db, location_id=id)
    if items:
        raise HTTPException(
            status_code=400, 
            detail="Cannot delete location with items. Move or delete items first."
        )
    
    # Check if location has children
    children = crud.location.get_multi_by_parent(db, parent_id=id, owner_id=current_user.id)
    if children:
        raise HTTPException(
            status_code=400, 
            detail="Cannot delete location with sub-locations. Delete sub-locations first."
        )
    
    location = crud.location.remove(db=db, id=id)
    return location 

File: frontend/src/services/locations.js
import api from './api';

export const getLocations = async (params = {}) => {
  try {
    const response = await api.get('/locations', { params });
    console.log('获取位置列表返回数据:', response.data);
    return response.data;
  } catch (error) {
    console.error('获取位置列表失败:', error);
    throw error;
  }
};

export const getLocationTree = async () => {
  try {
    const response = await api.get('/locations/tree');
    return response.data;
  } catch (error) {
    console.error('获取位置树失败:', error);
    throw error;
  }
};

export const getLocation = async (id) => {
  if (!id) {
    throw new Error('位置ID不能为空');
  }
  
  try {
    const response = await api.get(`/locations/${id}`);
    return response.data;
  } catch (error) {
    console.error(`获取位置详情失败 (ID: ${id}):`, error);
    throw error;
  }
};

export const createLocation = async (locationData) => {
  if (!locationData || !locationData.name) {
    throw new Error('位置名称不能为空');
  }
  
  // 映射字段名称，确保与API一致
  const apiPayload = {
    name: locationData.name,
    parent_id: locationData.parent_id,
    description: locationData.description,
    image_url: locationData.image_url
  };
  
  try {
    console.log('正在创建位置:', apiPayload);
    const response = await api.post('/locations', apiPayload);
    console.log('位置创建返回数据:', response.data);
    return response.data;
  } catch (error) {
    console.error('创建位置失败:', error, locationData);
    throw error;
  }
};

export const updateLocation = async (id, locationData) => {
  if (!id) {
    throw new Error('位置ID不能为空');
  }
  
  if (!locationData || !locationData.name) {
    throw new Error('位置名称不能为空');
  }
  
  // 映射字段名称，确保与API一致
  const apiPayload = {
    name: locationData.name,
    parent_id: locationData.parent_id,
    description: locationData.description,
    image_url: locationData.image_url
  };
  
  try {
    console.log(`正在更新位置(ID: ${id}):`, apiPayload);
    const response = await api.put(`/locations/${id}`, apiPayload);
    console.log(`位置更新返回数据(ID: ${id}):`, response.data);
    return response.data;
  } catch (error) {
    console.error(`更新位置失败 (ID: ${id}):`, error, locationData);
    throw error;
  }
};

export const deleteLocation = async (id) => {
  if (!id) {
    throw new Error('位置ID不能为空');
  }
  
  try {
    const response = await api.delete(`/locations/${id}`);
    return response.data;
  } catch (error) {
    console.error(`删除位置失败 (ID: ${id}):`, error);
    throw error;
  }
};

export const getLocationsByParent = async (parentId) => {
  try {
    const response = await api.get('/locations', { params: { parent_id: parentId } });
    return response.data;
  } catch (error) {
    console.error(`获取子位置失败 (父ID: ${parentId}):`, error);
    throw error;
  }
}; 